### PR TITLE
Implement getting forecast

### PR DESCRIPTION
**Pull request by ihhyy**

### CODE REVIEW

File: BL/Services/WeatherServices.cs [hunk 1]
```diff
@@ -17,6 +19,29 @@ public WeatherServices(IWeatherRepository weatherRepository, IValidator validato
             _validator = validator;
         }
 
+        public async Task<List<WeatherDto>> GetForecastByCityNameAsync(string cityName, int days)
+        {
+            _validator.ValidateInput(cityName);
```
Comment (denja2) [inline BL/Services/WeatherServices.cs]: Validate date amount

File: DAL/Entities/Weather.cs [hunk 2]
```diff
@@ -10,7 +8,6 @@ public class Weather
 
         public string Name { get; set; }
 
-        public int Code { get; set; }
-
+        public DateTime Dt_txt { get; set; }
```
Comment (denja2) [inline DAL/Entities/Weather.cs]: Use attributes

File: DAL/Repositories/WeatherRepository.cs [hunk 3]
```diff
@@ -1,34 +1,67 @@
...
     {
         private readonly HttpClient _client;
-        private readonly string _API;
+        private readonly string _currentWeatherUrl;
         private readonly string _key;
+        private readonly string _converterUrl;
+        private readonly string _forecastUrl;
 
-        public WeatherRepository(string key, string API, HttpClient client)
+        public WeatherRepository(string key, string currentWeatherUrl, HttpClient client, string converterUrl, string forecastUrl)
```
Comment (denja2) [inline DAL/Repositories/WeatherRepository.cs]: Unify params in a class

File: DAL/Repositories/WeatherRepository.cs [hunk 4]
```diff
@@ -1,34 +1,67 @@
...
         {
-            var response = await _client.GetAsync($"{_API}q={cityName}&appid={_key}&units=metric");
+            var response = await _client.GetAsync($"{_currentWeatherUrl}q={cityName}&appid={_key}&units=metric");
             var responseBody = await response.Content.ReadAsStringAsync();
             var weather = JsonConvert.DeserializeObject<Weather>(responseBody);
 
             return weather;
         }
+
+        public async Task<Forecast> GetWForecastByCityNameAsync(string cityName, int days)
```
Comment (denja2) [inline DAL/Repositories/WeatherRepository.cs]: Name

File: DAL/Repositories/WeatherRepository.cs [hunk 5]
```diff
@@ -1,34 +1,67 @@
...
+        private readonly string _forecastUrl;
 
-        public WeatherRepository(string key, string API, HttpClient client)
+        public WeatherRepository(string key, string currentWeatherUrl, HttpClient client, string converterUrl, string forecastUrl)
         {
             _client = client;
-            _API = API;
+            _currentWeatherUrl = currentWeatherUrl;
             _key = key;
+            _converterUrl = converterUrl;
```
Comment (denja2) [inline DAL/Repositories/WeatherRepository.cs]: coordinates

File: DAL/Repositories/WeatherRepository.cs [hunk 6]
```diff
@@ -1,34 +1,67 @@
...
 
             return weather;
         }
+
+        public async Task<Forecast> GetWForecastByCityNameAsync(string cityName, int days)
+        {
+            var cityCoord = await GetCoordinatesByCityName(cityName);
+            var weatherForecast = new Forecast();
+
+            if(cityCoord.Coord == null)
```
Comment (denja2) [inline DAL/Repositories/WeatherRepository.cs]: return null object

File: BL/Services/WeatherServices.cs [hunk 7]
```diff
@@ -17,6 +19,29 @@ public WeatherServices(IWeatherRepository weatherRepository, IValidator validato
...
+
+            var forecast = await _weatherRepositiry.GetWForecastByCityNameAsync(cityName, days);
+
+            if (forecast.IsBadRequest == true)
```
Comment (denja2) [inline BL/Services/WeatherServices.cs]: Add brackets

File: BL/Services/WeatherServices.cs [hunk 8]
```diff
@@ -17,6 +19,29 @@ public WeatherServices(IWeatherRepository weatherRepository, IValidator validato
...
+                return new List<WeatherDto>()
+                { 
+                    new WeatherDto {IsBadRequest = true, Message = "City not found or input was incorrect"}
+                };
+
+            var forecastList = new List<WeatherDto>();
+
+            foreach (var f in forecast.List.Where(x => x.Dt_txt.Hour == 12))
```
Comment (denja2) [inline BL/Services/WeatherServices.cs]: Move 12 to constant

Comment (denja2) [inline BL/Services/WeatherServices.cs]: Use Select

File: BL/Services/WeatherServices.cs [hunk 9]
```diff
@@ -17,6 +19,29 @@ public WeatherServices(IWeatherRepository weatherRepository, IValidator validato
             _validator = validator;
         }
 
+        public async Task<List<WeatherDto>> GetForecastByCityNameAsync(string cityName, int days)
```
Comment (denja2) [inline BL/Services/WeatherServices.cs]: Return WeatherDto with full string

**Review by denja2 — CHANGES_REQUESTED** [review]

File: DAL/Entities/Weather.cs (hunk 2 above)
Comment (ihhyy) [reply DAL/Entities/Weather.cs]: Done

File: DAL/Repositories/WeatherRepository.cs (hunk 4 above)
Comment (ihhyy) [reply DAL/Repositories/WeatherRepository.cs]: Fixed

File: BL/Services/WeatherServices.cs (hunk 1 above)
Comment (ihhyy) [reply BL/Services/WeatherServices.cs]: Done

File: DAL/Repositories/WeatherRepository.cs (hunk 5 above)
Comment (ihhyy) [reply DAL/Repositories/WeatherRepository.cs]: Renamed

File: BL/Services/WeatherServices.cs (hunk 7 above)
Comment (ihhyy) [reply BL/Services/WeatherServices.cs]: Done

File: DAL/Repositories/WeatherRepository.cs (hunk 3 above)
Comment (ihhyy) [reply DAL/Repositories/WeatherRepository.cs]: Done

File: BL/Services/WeatherServices.cs (hunk 8 above)
Comment (ihhyy) [reply BL/Services/WeatherServices.cs]: Done

Comment (ihhyy) [reply BL/Services/WeatherServices.cs]: Done

File: AppConfig/Config.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+using System.Threading.Tasks;
+using AppConfig.Interfaces;
+using System.Configuration;
+
+namespace AppConfig
+{
+    public class Config : IConfig
+    {
+        public string Key => ConfigurationManager.AppSettings["APIKey"];
+
```
Comment (Xodoc) [inline AppConfig/Config.cs:17]: Move it to an extension method or read the config in the constructor and set the properties through the indexer

File: BL/Services/WeatherInputValidator.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            if (string.IsNullOrEmpty(input))
+                throw new EmptyInputException();
+        }
+
+        public void ValidateMultiInput(string input, int days)
+        {
+            var min = int.Parse(_config.MinDays);
+            var max = int.Parse(_config.MaxDays);
+
+            if (days == 0)
```
Comment (Xodoc) [inline BL/Services/WeatherInputValidator.cs]: days == min

File: BL/Services/WeatherServices.cs [hunk 12]
```diff
@@ -3,18 +3,43 @@
...
+        public async Task<string> GetForecastByCityNameAsync(string cityName, int days)
+        {
+            _validator.ValidateMultiInput(cityName, days);
+
+            var forecast = await _weatherRepositiry.GetForecastByCityNameAsync(cityName);
+
+            if (forecast.IsBadRequest)
+                return "City not found or input was incorrect";
+
+            var forecastList = forecast.List.Where(x => x.Date.Hour == _forecastHour)
```
Comment (Xodoc) [inline BL/Services/WeatherServices.cs:32]: Thinking it should be in the DAL

File: Command/Commands/GetCurrentWeatherCommand.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        private readonly string _coordinatesUrl;
+        private IWeatherService _weatherService;
+        private IWeatherRepository _weatherRepository;
+        private IValidator _validator;
+
+        public string Text => ": Get current weather";
+
+        public GetCurrentWeatherCommand()
+        {
+            _key = _config.Key;
```
Comment (Xodoc) [inline Command/Commands/GetCurrentWeatherCommand.cs]: Memory is allocated for unnecessary variables. It's better to pass value to WeatherRepository via _config

File: Command/Commands/GetWeatherForecastCommand.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Command.Commands
+{
+    public class GetWeatherForecastCommand : ICommand
+    {
+        private static IConfig _config = new Config();
```
Comment (Xodoc) [inline Command/Commands/GetWeatherForecastCommand.cs]: Remove "= new Config()" and write it "_config = new Config()" in a GetWeatherForecastCommand

File: Config/Configuration.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+﻿using Config.Interfaces;
+using System.Configuration;
+
+namespace Config
+{
+    public class Configuration : IConfig
+    {
+        public string Key => ConfigurationManager.AppSettings["APIKey"];
```
Comment (Xodoc) [inline Config/Configuration.cs:8]: .

File: DAL/Entities/Forecast.cs [hunk 16]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (Xodoc) [inline DAL/Entities/Forecast.cs:3]: Remove unnecessary namespace

File: IntegrationTests/Tests/GetWeatherByCityNameIntegrationTests.cs [hunk 17]
```diff
@@ -20,16 +20,20 @@ public class GetWeatherByCityNameIntegrationTests
         private readonly IValidator _validator;
         private readonly IWeatherService _weatherService;
         private readonly string _key;
```
Comment (Xodoc) [inline IntegrationTests/Tests/GetWeatherByCityNameIntegrationTests.cs:22]: .

File: BL/Services/WeatherInputValidator.cs (hunk 11 above)
Comment (denja2) [reply BL/Services/WeatherInputValidator.cs]: +

File: BL/Services/WeatherInputValidator.cs [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        public void ValidateMultiInput(string input, int days)
+        {
+            if (days == 0)
+                throw new EmptyInputException();
+
+            if (days < _min || days > _max)
+                throw new IncorrectDaysRangeException();
+
+
+            this.ValidateInput(input);
```
Comment (denja2) [inline BL/Services/WeatherInputValidator.cs]: no need to use this keyword

File: AppConfig/Config.cs (hunk 10 above)
Comment (ihhyy) [reply AppConfig/Config.cs:17]: Done

File: Command/Commands/GetWeatherForecastCommand.cs (hunk 14 above)
Comment (ihhyy) [reply Command/Commands/GetWeatherForecastCommand.cs]: Fixed

File: Config/Configuration.cs (hunk 15 above)
Comment (ihhyy) [reply Config/Configuration.cs:8]: Fixed

File: DAL/Entities/Forecast.cs (hunk 16 above)
Comment (ihhyy) [reply DAL/Entities/Forecast.cs:3]: Removed

File: AppConfig/Config.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+        public string CurrentWeatherUrl { get; set; }
+        public string CoordinatesUrl { get; set; }
+        public string ForecastUrl { get; set; }
+        public string ForecastHour { get; set; }
+        public string MinDays { get; set; }
+        public string MaxDays { get; set; }
+
+        public Config()
+        {
+            var configuration = Configuration.InitConfig();
```
Comment (denja2) [inline AppConfig/Config.cs:20]: See Alessia's approach

File: AppConfig/Config.cs [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+using System.Configuration;
+
+namespace AppConfig
+{
+    public class Config : IConfig
+    {
+        public readonly NameValueCollection Configuration;
+        //public string Key => ConfigurationManager.AppSettings["APIKey"];
+
+        //public string CurrentWeatherUrl => ConfigurationManager.AppSettings["currentWeatherUrl"];
```
Comment (denja2) [inline AppConfig/Config.cs]: remove comments

File: AppConfig/Interfaces/IConfig.cs [hunk 21]
```diff
@@ -0,0 +1,19 @@
+﻿namespace AppConfig.Interfaces
+{
+    public interface IConfig
```
Comment (denja2) [inline AppConfig/Interfaces/IConfig.cs:3]: No reason to have the interface

File: BL/Services/WeatherInputValidator.cs [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+        public void ValidateInput(string input)
+        {
+            if (string.IsNullOrEmpty(input))
+                throw new EmptyInputException();
+        }
+
+        public void ValidateMultiInput(string input, int days)
+        {
+            if (days == 0)
```
Comment (denja2) [inline BL/Services/WeatherInputValidator.cs]: what is it? why resolved without comment

File: BL/Services/WeatherServices.cs [hunk 23]
```diff
@@ -3,18 +3,43 @@
...
+                .Select(x => MapEntityToWeatherDto(x, forecast.City.Name)).ToList();
+
+            string fullMessage = string.Empty;
+
+            for(int i = 0; i < days; i++)
+            {
+                fullMessage += SelectPrefix(i) + forecastList[i].Message + "\n";
```
Comment (denja2) [inline BL/Services/WeatherServices.cs]: better to use interpolation

File: Command/Commands/GetCurrentWeatherCommand.cs [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+using DAL.Repositories;
+using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Command.Commands
+{
+    public class GetCurrentWeatherCommand : ICommand
+    {
+        private static IConfig _config = new Config();
```
Comment (denja2) [inline Command/Commands/GetCurrentWeatherCommand.cs]: Initialize through constructor

File: AppConfig/Config.cs [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+
+        //public string MinDays => ConfigurationManager.AppSettings["minForecastDays"];
+
+        //public string MaxDays => ConfigurationManager.AppSettings["maxForecastDays"];
+        public string Key { get; set; }
+        public string CurrentWeatherUrl { get; set; }
+        public string CoordinatesUrl { get; set; }
+        public string ForecastUrl { get; set; }
+        public string ForecastHour { get; set; }
+        public string MinDays { get; set; }
```
Comment (denja2) [inline AppConfig/Config.cs]: Numeric convert to int

File: Config/Interfaces/IConfig.cs [hunk 26]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Config.Interfaces
+{
+    public interface IConfig
```
Comment (denja2) [inline Config/Interfaces/IConfig.cs:7]: Unify config

File: ConsoleApp/Program.cs [hunk 27]
```diff
@@ -1,73 +1,72 @@
...
+
+            var list = new List<ICommand>() 
+            {
+                exitCommand, currentWeatherCommand, forecastCommand
+            };
+
             while (showMenu)
             {
-                showMenu = await MainMenu();
+                showMenu = await MainMenu(list);
```
Comment (denja2) [inline ConsoleApp/Program.cs]: Method name should be a verb

File: ConsoleApp/Program.cs [hunk 28]
```diff
@@ -1,73 +1,72 @@
...
-                    return false;
-                default:
+                Console.WriteLine("Chose an option:");
+
+                for (int i = 0; i < list.Count; i++)
+                {
+                    Console.WriteLine(i + list[i].Text);
+                }
+
+                var selectedNumber = int.Parse(Console.ReadLine());
```
Comment (denja2) [inline ConsoleApp/Program.cs]: TryParse

File: DAL/Repositories/WeatherRepository.cs [hunk 29]
```diff
@@ -1,34 +1,69 @@
...
+            var cityCoord = await GetCoordinatesByCityName(cityName);
+            var weatherForecast = new Forecast();
+
+            if(cityCoord == null)
+            {
+                weatherForecast.IsBadRequest = true;
+                return weatherForecast;
+            }
+            else
+            {
```
Comment (denja2) [inline DAL/Repositories/WeatherRepository.cs]: ?

File: BL/Services/WeatherInputValidator.cs (hunk 18 above)
Comment (ihhyy) [reply BL/Services/WeatherInputValidator.cs]: Removed

File: AppConfig/Config.cs (hunk 20 above)
Comment (ihhyy) [reply AppConfig/Config.cs]: Removed

File: DAL/Repositories/WeatherRepository.cs (hunk 29 above)
Comment (ihhyy) [reply DAL/Repositories/WeatherRepository.cs]: Fixed

File: ConsoleApp/Program.cs (hunk 28 above)
Comment (ihhyy) [reply ConsoleApp/Program.cs]: Remade

File: ConsoleApp/Program.cs (hunk 27 above)
Comment (ihhyy) [reply ConsoleApp/Program.cs]: Renamed

File: AppConfig/Config.cs (hunk 25 above)
Comment (ihhyy) [reply AppConfig/Config.cs]: Fixed

File: BL/Services/WeatherInputValidator.cs (hunk 22 above)
Comment (ihhyy) [reply BL/Services/WeatherInputValidator.cs]: Fixed

File: BL/Services/WeatherServices.cs (hunk 23 above)
Comment (ihhyy) [reply BL/Services/WeatherServices.cs]: Done

File: Command/Commands/GetCurrentWeatherCommand.cs (hunk 24 above)
Comment (ihhyy) [reply Command/Commands/GetCurrentWeatherCommand.cs]: Done

**Pull request merged**